fix(autonomi): verify register on fetch

b-zee opened the pull request:
Registers are fetched and contain CRDT operations. These operations
should be valid, so we check once we fetched it. Then later we assert
the register is valid.

davidrusu commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -101,39 +125,44 @@ impl SectionPeers {
 
-            updated |= match (self.members.entry(node_name), new_state.state()) {
-                (Entry::Vacant(entry), MembershipState::Joined) => {
+            updated |= match (self.members().contains(new_state), new_state.state()) {
+                (false, MembershipState::Joined) => {
                     // unless it was already archived, insert it as current member
                     if self.archive.contains_key(&node_name) {
                         false
                     } else {
-                        entry.insert(new_decision.clone());
+                        shall_push_to_decisions = true;
                         true
                     }
                 }
-                (Entry::Vacant(_), MembershipState::Left | MembershipState::Relocated(_)) => {
+                (false, MembershipState::Left | MembershipState::Relocated(_)) => {
                     // insert it in our archive regardless it was there with another state
                     let _prev = self.archive.insert(node_name, new_decision.clone());
                     true
                 }
-                (Entry::Occupied(_), MembershipState::Joined) => false,
-                (Entry::Occupied(entry), MembershipState::Left | MembershipState::Relocated(_)) => {
-                    //  remove it from our current members, and insert it into our archive
-                    let _ = entry.remove();
+                (true, MembershipState::Joined) => false,
+                (true, MembershipState::Left | MembershipState::Relocated(_)) => {
+                    shall_push_to_decisions = true;
                     let _ = self.archive.insert(node_name, new_decision.clone());
                     true
                 }
             };
         }
+
+        if shall_push_to_decisions {
+            self.decisions.push(new_decision);
+        }
+
         updated
     }
```
Decisions need to be processed in order by generation, so you can reduce this function to something like:

```rust
pub(super) fn update(&mut self, decision: Decision<NodeState>) -> Result<bool> {
    if new_decision.generation()? == self.decisions.len() + 1 {
        self.decisions.push(new_decision);

        for archived_node in decision.proposals.iter().filter(|n| matches!(n.state(), MembershipState::Left)) {
           self.archive.insert(archived_node.name(), decision.clone());
        }

        Ok(true)
    } else {
        // TODO: request AE update with previous decisions.
        Ok(false)
    }
}
```

To make this work, I think we may need to also change the `AntiEntropyKind::Update(..)` method to be `AntiEntropyKind::Update(Vec<Decision>)`

maqi replied:
> To make this work, I think we may need to also change the AntiEntropyKind::Update(..) method to be AntiEntropyKind::Update(Vec<Decision>)

The current Update sends `SectionPeers` which contains `Vec<Decision>` already. 
We can have a later on PR to optimisze it, but currently it shall be OK as long as the included content got properly applied.

davidrusu replied:
Confusingly, when I traced the `SectionPeers` type, it was actually a type alias of `BTreeSet<SectionSig<NodeState>`.

maqi replied:
It's an alias of `BTreeSet<Decision<NodeState>>;` actually ?
I will double check maybe `Vec<Decision<NodeState>>` will be better, might rename it as well.
OR, I might just replace it with the proper `SectionPeers` to be used.

davidrusu replied:
I think just inlining `Vec<Decision<NodeState>>` into the Update message is good enough

oetyng commented on sn_interface/src/network_knowledge/errors.rs:
```diff
@@ -68,4 +68,6 @@ pub enum Error {
     #[error("An archived node attempted to rejoin the section")]
     ArchivedNodeRejoined,
+    #[error("Request AE update")]
+    RequestAEUpdate,
```
What is the actual error here?
Requesting an AE update isn't an error?

maqi replied:
we are behind hence request an AE update, I will expand the display msg to make it more explicit

oetyng replied:
`AEOutdated` is the actual error then.
`RequestAEUpdate` is an imperative (something we must do, i.e. a cmd) resulting from us being outdated.

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -200,17 +189,14 @@ impl NetworkKnowledge {
-                    sig,
-                }));
-        }
+        let bootstrap_members = network_knowledge
```
We only expect a single Elder here, right? the first node.

maqi replied:
yes. it's just the API is like that.

oetyng replied:
Maybe we can just pick that single elder, and make it clear in the naming that we only have and expect one?

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -200,17 +189,14 @@ impl NetworkKnowledge {
+            .elders()
+            .map(|peer| NodeState::joined(*peer, None))
+            .collect();
+        network_knowledge
```
What's happening here?

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -427,51 +422,47 @@ impl NetworkKnowledge {
-        self.section_peers.retain(&self.prefix());
-
-        Ok(there_was_an_update)
+    /// This function shall only get called at the time of SAP change
```
Also good for docs: _Why_ is it only called at the time of SAP change, and what does it mean in the domain to reset the bootstrap members?

maqi replied:
> Why is it only called at the time of SAP change

The member changes between the SAP change will be recorded via updates, i.e. stored Decisions

> bootstrap members

Any idea of a better naming?

oetyng replied:
Yeah I think `bootstrap` is a concept we have specifically for joining a network, and it makes things confusing to reuse the word for other things.

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -427,51 +422,47 @@ impl NetworkKnowledge {
+    }
+
+    /// Update one decision. Returns whether it actually updated.
+    pub fn update_decision(&mut self, decision: Decision<NodeState>) -> Result<bool> {
```
I thought by the name that the decision itself was updated, but there was no mut ref and nothing updated.

So this is a check telling us if we need to update a decision?

Name perhaps: `shall_update_decision` or `decision_update_needed` ?

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -427,51 +422,47 @@ impl NetworkKnowledge {
+        Ok(self.section_peers.update_peers(peers))
+    }
+
+    /// Update one decision. Returns whether it actually updated.
```
Are we actually updating a decision?

It seems we are updating our section peers with a new decision?

maqi replied:
Yeah some wording need to be optimized, I will change the naming bit.

oetyng commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -520,9 +511,9 @@ impl NetworkKnowledge {
-    /// Returns joined members with Decision info.
-    pub fn section_members_with_decision(&self) -> BTreeSet<Decision<NodeState>> {
-        self.section_peers.section_members_with_decision()
+    /// Returns joined members at the specific generation.
```
Nit pick: Aren't members just simply members?
So when a node has left, it is not a member.

I don't think we need an extra qualifier like _joined_ beause that introduces doubts on what exactly "member" means. Can a member not be a part of the section?

maqi replied:
I think it's just to highlight, as there we have `archived_members`
As `member` within `SectionPeers` also have the meaning of `once being a member`

oetyng replied:
Maybe `archived` is simpler and more to the point if just being called `previous_members`. That way there is no way to explain that `archived` means they were previously members.

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,74 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
 pub(super) struct SectionPeers {
-    members: BTreeMap<XorName, Decision<NodeState>>,
+    /// Initial members snapshot at the time of SAP change
+    bootstrap_members: BTreeSet<NodeState>,
```
Could we call this `initial_members` instead?

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,74 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
+    bootstrap_members: BTreeSet<NodeState>,
+    /// Decisions received during the current SAP so far
+    decisions: Vec<Decision<NodeState>>,
+    /// Left Decisions archived
```
Are these archived decisions on leaves only?

I think the comment and field can be clearer on that.

(Perhaps also some comment on why we archive leaves specifically and nothing else.)

maqi replied:
To be more accurate, it's actually `decision contains a leaving node_state (left or relocated)` to be archived.
The reason is to avoid node's rejoin using the same id.
I will update the comment little bit

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,74 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
 /// Container for storing information about (current and archived) members of our section.
-#[derive(Clone, Default, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Default, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
 pub(super) struct SectionPeers {
```
Note: 
There seems to be some more domain logic expressed with `SectionPeers` now.
`SectionPeers` seems to be an instance that is reflecting the span of a single SAP, continuously updated with Adult changes.
I'm thinking of how that can be made clearer with the naming.

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,74 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
 
 impl SectionPeers {
+    /// Reset the bootstrap members when SAP changed.
+    pub(crate) fn reset_bootstrap_members(&mut self, new_bootstrap_members: BTreeSet<NodeState>) {
```
Okay, I can see now from reading though more code (considering my previous questions on meaning of this fn), that this is a re-instantiation of section peers.

I think it would actually be better for clarity to just replace the instance.
Like so:
`self.section_peers = SectionPeers::new(initial_elders);`

oetyng replied:
Alternatively rename the reset to:

`transition_to_new_elders` (because those are always Elders, right? Never Adults?)

maqi replied:
> to just replace the instance

We still need the `archived_members` from the previous SectionPeer

> rename the reset to:  transition_to_new_elders (because those are always Elders, right? Never Adults?)

No. at the time of the SAP change, all members (elders + adults) at that time becomes `bootstrap_emmbers/initial_members`

oetyng replied:
I think it would be more specific and clear what we want and are doing if we call a fn `update_members` for example, and pass in the new set `members`.

(there is some elder change, and there could be an adult change as well if it was due to promotion)

I think that would make it all much clearer.

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,74 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
+        self.decisions.clone()
+    }
+
+    /// Returns joined members at the specific generation.
```
Same as previous comment about members and `joined` qualifier.

oetyng replied:
Members are just members. Other nodes are not members.
The impl details is that we identify them as those that saw the event Joined but not the event Left or Relocated.

(Now, `NodeState` is actually in our domain a `NodeEvent` (`Left`, `Joined`, `Relocated` are _past tense_, things that _happened_, i.e. events), and a `NodeState` would be a left fold of such events (replaying them to get current state), giving for example `NodeState::Member`, `NodeState::PreviousMember`, `NodeState::Other`).

maqi replied:
https://github.com/maidsafe/safe_network/pull/2124#discussion_r1119911069

oetyng commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -293,15 +350,17 @@ mod tests {
+        assert!(!section_peers.update(node_left_joins)?);
+        assert!(!section_peers.update(node_relocated_joins)?);
 
         assert_lists(section_peers.archive.values(), &[node_left, node_relocated]);
```
Wrt the comment on the `archive` field of `SectionPeers` (saying `/// Left Decisions archived`), this code here seems to imply that the archive contains the events of nodes `joined` decided, but we put them in `archive` after we have seen the events of them `left` decided as well as `relocated` decided.

maqi replied:
`archive` is a btreemap with key of node_name, i.e. although the Decision may contain multiple node_states, the entry is only about one node.
and we only interested in the `node_state of that entry's node_name within the Decision` to be left.

oetyng replied:
What I mean is that the archive actually contains `Joined` states (that's what we are comparing against in the test, and are adding to our set for comparison).

So here in the test, it looks like the state is supposed to be (or can be, nothing clear) `Joined`in the archive.

But in the comment on the archive field (and your comments here?) seem to say they should only be `Left`?

maqi replied:
these two lines of assertion no long valid, and there is now a TODO comment put in to explain the reason.

> the archive actually contains Joined states 

It will only happen when a Decision holds multiple node_states, and one of it is a `Left` or `Relocated` . 
Also, for `archive` , for an entry's decision value, we only interested in the node_state that its name matches the entry's name.

davidrusu commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -198,17 +187,10 @@ impl NetworkKnowledge {
-                    sig,
-                }));
-        }
+        let initial_members: BTreeSet<_> = iter::once(NodeState::joined(peer, None)).collect();
```
I find the `from_iter(..)` pattern more clear:
```suggestion
        let initial_members = BTreeSet::from_iter([NodeState::joined(peer, None)]);
```
Either way works 👍

maqi replied:
`NodeState::joined(peer, None)` is not an iter, hence has to call some `into_iter` if want to using `from_iter` .
so still prefer to use `: BTreeSet<_>` in this case.

davidrusu commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -19,33 +23,68 @@ const ELDER_CHURN_EVENTS_TO_PRUNE_ARCHIVE: usize = 5;
 
 /// Container for storing information about (current and archived) members of our section.
-#[derive(Clone, Default, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Default, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
```
I don't think we need the `Serialize, Deserialize` derives anymore since we are no longer including the full SectionPeers in the AEupdate

davidrusu commented on sn_interface/src/network_knowledge/section_peers.rs:
```diff
@@ -78,45 +117,65 @@ impl SectionPeers {
     }
 
-    /// Remove all members whose name does not match `prefix`.
+    pub(super) fn update_peers(&mut self, peers: Vec<Decision<NodeState>>) -> bool {
+        let mut there_was_an_update = false;
+
+        // TODO: validate all the incoming Decisions are signed by current SAP
+        if self.decisions.len() < peers.len() {
+            // Only update to the newer
+            self.decisions = peers;
+            there_was_an_update = true;
+        }
+
+        there_was_an_update
+    }
```
From slack convo, we want to remove this function since we only want one way to update the section peers decisions.

davidrusu commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -330,7 +318,9 @@ impl MyNode {
-        let members = self.get_members_at_gen(sap.membership_gen())?;
+
+        // use the current members
+        let members = self.get_current_members()?;
```
This will cause Handover failures which will lead to failures in elder promotions, you need to use the members at the time that the Sap candidates were created. (i.e. revert back to `members_at_gen(sap.membership_gen())` logic)

davidrusu commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -356,7 +346,7 @@ impl MyNode {
         let prev_prefix = sap1.prefix().popped();
         let previous_gen_sap = self.get_sap_for_prefix(prev_prefix)?;
-        let members = self.get_members_at_gen(sap1.membership_gen())?;
+        let members = self.get_current_members()?;
```
same here.

b-zee committed:
fix(autonomi): verify register on fetch

Registers are fetched and contain CRDT operations. These operations
should be valid, so we check once we fetched it. Then later we assert
the register is valid.

joshuef commented:
there was a bug that led you here right @b-zee ? Can we get a test case of that bug?

(also what's the implication of the change from `base_register()` to `register()` ?

b-zee commented:
>(also what's the implication of the change from base_register() to register() ?

This was the source of the bug. I assumed when we freshly fetched a register, the 'base register' would contain all the values, but this is only after applying the CRDT operations. Getting that base register plus operations was a fallible method, which was why I didn't like calling it.

maqi commented:
> when we freshly fetched a register, the 'base register' would contain all the values, but this is only after applying the CRDT operations.

Yeah, that's true.
I think this might being the reason why it is called `base_register` within `SignedRegister`.
I do agree the current naming/flow is kind of `confusing` (as I am also just started touch this part due to recent community member's issue raised related to it)

However, I think it is kind of `mis-used`, as `ops` shall always just being `local mutation history`, and published `register record` shall only retain the `MerkleReg` for the `last updated value + DAG tree for the update history` ?

Following is what I think shall happen to `ops` and `base_register` member fields of `SignedRegister` : 
* the `ops` are all `local` ones to record `mutate instructions issued by client`
* client locally write `mutate instructions` into `SignedRegister.ops`
* during `publish`, all `ops` will be applied to the `base_register` and then uploaded to network (and only `base_register` got uploaded as a `register record`)
* When fetch from network, a signed_register using the fetched register record as its base_register will be created, with its `ops` to be empty
